src/m365/spo/commands/roledefinition/roledefinition-add.spec.ts

```diff
@@ -0,0 +1,154 @@
+  let logger: Logger;
+  let commandInfo: CommandInfo;
+
+  before(() => {
+    sinon.stub(auth, 'restoreAuth').callsFake(() => Promise.resolve());
+    sinon.stub(appInsights, 'trackEvent').callsFake(() => { });
+    auth.service.connected = true;
+    commandInfo = Cli.getCommandInfo(command);
+  });
+
+  beforeEach(() => {
+    log = [];
+    logger = {
+      log: (msg: string) => {
+        log.push(msg);
+      },
+      logRaw: (msg: string) => {
+        log.push(msg);
+      },
+      logToStderr: (msg: string) => {
+        log.push(msg);
+      }
+    };
+  });
+
+  afterEach(() => {
+    sinonUtil.restore([
+      request.post
+    ]);
+  });
+
+  after(() => {
+    sinonUtil.restore([
+      auth.restoreAuth,
+      appInsights.trackEvent
+    ]);
+    auth.service.connected = false;
+  });
+
+  it('has correct name', () => {
+    assert.strictEqual(command.name.startsWith(commands.ROLEDEFINITION_ADD), true);
+  });
+
+  it('has a description', () => {
+    assert.notStrictEqual(command.description, null);
+  });
+
+  it('supports debug mode', () => {
+    const options = command.options;
+    let containsDebugOption = false;
+    options.forEach(o => {
+      if (o.option === '--debug') {
+        containsDebugOption = true;
+      }
+    });
+    assert(containsDebugOption);
+  });
+
+  it('fails validation if the webUrl option is not a valid SharePoint site URL', async () => {
+    const actual = await command.validate({ options: { webUrl: 'foo', name: 'abc' } }, commandInfo);
+    assert.notStrictEqual(actual, true);
+  });
+
+  it('passes validation if the webUrl option is a valid SharePoint site URL', async () => {
+    const actual = await command.validate({ options: { webUrl: 'https://contoso.sharepoint.com', name: 'abc' } }, commandInfo);
+    assert.strictEqual(actual, true);
+  });
+
+  it('fails if non existing PermissionKind rights specified', async () => {
+    const actual = await command.validate({ options: { webUrl: 'https://contoso.sharepoint.com', name: 'abc', rights: 'abc' } }, commandInfo);
+    assert.strictEqual(actual, `Rights option 'abc' is not recognized as valid PermissionKind choice. Please note it is case sensitive`);
+  });
+
+  it('has correct PermissionKind rights specified', async () => {
+    const actual = await command.validate({ options: { webUrl: 'https://contoso.sharepoint.com', name: 'abc', rights: 'FullMask' } }, commandInfo);
+    assert.strictEqual(actual, true);
+  });
+
+  it('offers autocomplete for the rights option', () => {
+    const options = command.options;
+    for (let i = 0; i < options.length; i++) {
+      if (options[i].option.indexOf('--rights') > -1) {
+        assert(options[i].autocomplete);
+        return;
+      }
+    }
+    assert(false);
+  });
+
+  it('adds role definition to web with name, description and right', (done) => {
+    sinon.stub(request, 'post').callsFake((opts) => {
+      if ((opts.url as string).indexOf('/_api/web/roledefinitions') > -1) {
```

Comment:
Also something what I try to pay attention to is not to use `.indexOf(...)` but specify the entire URL.
`opts.url === 'https;//contoso.sharepoint.com/_api/web/roledefinitions'`
In the past we've had a refactor happening where an unknown option was passed in the request logic. This resulted in an (let's say) `undefined/_api/web/roledefinitions` request which kind of broke the command. All test used this `.indexOf(...)` kind of approach and all succeeded.
If we specify the entire URL, this test would fail and makes sure this mistake cannot happen again 😃

Reply:
done. I would need to start following this approach in other places as well. thanks for this comment 👍